File: rateLimiter/src/ratelimit/strategy/SlidingWindowCounter.java
package ratelimit.strategy;

import java.time.Duration;
import java.time.Instant;

import ratelimit.model.RateLimitDecision;

public class SlidingWindowCounter implements RateLimitStrategy {
    private final int limit;
    private final Duration window;
    private int currentCount;
    private Instant currentStart;
    private int prevCount;

    public SlidingWindowCounter(int limit, Duration window, Instant start) {
        this.limit = limit;
        this.window = window;
        this.currentStart = start;
        this.currentCount = 0;
        this.prevCount = 0;
    }

    @Override
    public synchronized RateLimitDecision tryAcquire(Instant reqTime) {
        Instant currentEnd = currentStart.plus(window);
        // Snap to aligned windows — do not set currentStart = reqTime (same bug as early FixedWindow).
        // Request at 12:01:30 with a 1min window belongs in [12:01, 12:02), not a new window at 12:01:30.
        if (!reqTime.isBefore(currentEnd)) {
            long windowsPassed = Duration.between(currentStart, reqTime).dividedBy(window);
            if (windowsPassed == 1) {
                prevCount = currentCount; // adjacent window: old current becomes previous
            } else {
                prevCount = 0; // skipped ≥2 windows; the gap had no traffic
            }
            currentStart = currentStart.plus(window.multipliedBy(windowsPassed));
            currentCount = 0;
            currentEnd = currentStart.plus(window);
        }

        // Sliding view [now - W, now] still covers `overlap` of the previous fixed window.
        // At elapsed=0, weight=100% of prev. At elapsed=W, weight=0.
        long windowNanos = window.toNanos();
        long elapsed = Duration.between(currentStart, reqTime).toNanos();
        long overlap = windowNanos - elapsed;
        int weightedPrev = (int) ((overlap * (long) prevCount) / windowNanos);
        int estimated = weightedPrev + currentCount;

        if (estimated < limit) {
            currentCount++;
            // estimated is before this request; remaining after taking one is limit - estimated - 1
            return new RateLimitDecision(limit - estimated - 1, true, null);
        }
        // Approximate retry: when this fixed window ends, previous weight drops to 0.
        return new RateLimitDecision(0, false, currentEnd);
    }
}

File: rateLimiter/infra/redis/lua/sliding_window_counter.lua
-- Atomic sliding-window counter (matches Java ratelimit.strategy.SlidingWindowCounter).
--
-- KEYS[1]  key
-- ARGV[1]  limit
-- ARGV[2]  window_ms
-- ARGV[3]  now_ms        — optional epoch ms (tests); default Redis TIME
--
-- Returns: { allowed, remaining, retry_at_ms }
--   retry_at_ms = current window end when denied (approximate); -1 when allowed
-- TTL: 2 windows — previous count is still needed for the weight.

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window_ms = tonumber(ARGV[2])

local now_ms
if ARGV[3] ~= nil and ARGV[3] ~= '' then
  now_ms = tonumber(ARGV[3])
else
  local t = redis.call('TIME')
  now_ms = tonumber(t[1]) * 1000 + math.floor(tonumber(t[2]) / 1000)
end

local current_start = tonumber(redis.call('HGET', key, 'current_start_ms'))
local current_count = tonumber(redis.call('HGET', key, 'current_count'))
local prev_count = tonumber(redis.call('HGET', key, 'prev_count'))

if current_start == nil then
  current_start = now_ms
  current_count = 0
  prev_count = 0
end

local current_end = current_start + window_ms
if now_ms >= current_end then
  local windows_passed = math.floor((now_ms - current_start) / window_ms)
  if windows_passed < 1 then
    windows_passed = 1
  end
  if windows_passed == 1 then
    prev_count = current_count
  else
    prev_count = 0
  end
  current_start = current_start + windows_passed * window_ms
  current_count = 0
  current_end = current_start + window_ms
end

local elapsed = now_ms - current_start
local overlap = window_ms - elapsed
if overlap < 0 then
  overlap = 0
end
local weighted_prev = math.floor((overlap * prev_count) / window_ms)
local estimated = weighted_prev + current_count

local allowed
local remaining
local retry_at_ms

if estimated < limit then
  current_count = current_count + 1
  allowed = 1
  remaining = limit - estimated - 1
  retry_at_ms = -1
else
  allowed = 0
  remaining = 0
  retry_at_ms = current_end
end

redis.call('HSET', key,
  'current_start_ms', current_start,
  'current_count', current_count,
  'prev_count', prev_count)
redis.call('PEXPIRE', key, math.floor(2 * window_ms))

return { allowed, remaining, retry_at_ms }
